### PR TITLE
feat(hyperliquid) - watchTickers

### LINKED ISSUE #22927: Hyperliquid fetchticker/fetchtickers
### Preliminary Checks

- [X] I have already searched for [existing issues](https://github.com/ccxt/ccxt/issues) and confirmed that this issue is not a duplicate

### Is your feature request related to a problem? Please describe

Hello, in hyperliquid, fetchticker/fetchtickers is not supported. Freqtrade seems to require it. Any plans to implement fetchticker/fetchtickers?

### Describe the solution you'd like

Implementation of fetchticker/fetchtickers

### Describe alternatives you've considered

I don't think there is a workaround

### PR DESCRIPTION
**Pull request by ttodua**
this subscription is technically similar to other subscriptions, just this sub's request params are not documented, however, these are what is used to show tickers and works for now to fix for users like #22927. idk if this is not ideal solution, but i think it would work before they release the official watch-tickers in api.

```
npm run cli.ts hyperliquid watchTickers

> ccxt@4.3.62 cli.ts
> tsx examples/ts/cli.ts hyperliquid watchTickers

2024-07-16T19:05:22.782Z
Node.js: v21.6.2
CCXT v4.3.62
hyperliquid.watchTickers ()

'ETHFI:USDC/USDC': {
    symbol: 'ETHFI:USDC/USDC',
    previousClose: 2.245,
    close: 2.3298,
    bid: 2.32863,
    ask: 2.33069,
    quoteVolume: 886548.20994,
    info: {
      funding: '-0.00009416',
      openInterest: '992217.2',
      prevDayPx: '2.245',
      dayNtlVlm: '886548.20994',
      premium: '-0.0012042',
      oraclePx: '2.3335',
      markPx: '2.3295',
      midPx: '2.3298',
      impactPxs: [ '2.32863', '2.33069' ],
      coin: 'ETHFI:USDC/USDC'
    },
    bidVolume: undefined,
    askVolume: undefined,
    high: undefined,
    low: undefined,
    open: undefined,
    last: 2.3298,
    change: undefined,
    percentage: undefined,
    average: undefined,
    vwap: undefined,
    baseVolume: undefined
  }


'PEPE/USDC': {
    symbol: 'PEPE/USDC',
    previousClose: 0.001649,
    close: 0.001587,
    bid: undefined,
    ask: undefined,
    quoteVolume: 999.346969,
    info: {
      prevDayPx: '0.001649',
      dayNtlVlm: '999.346969',
      markPx: '0.001969',
      midPx: '0.001587',
      circulatingSupply: '99952395.828724',
      coin: '@11'
    },
    bidVolume: undefined,
    askVolume: undefined,
    high: undefined,
    low: undefined,
    open: undefined,
    last: 0.001587,
    change: undefined,
    percentage: undefined,
    average: undefined,
    vwap: undefined,
    baseVolume: undefined
  },
```

[close #22927]

### CODE REVIEW

File: ts/src/pro/hyperliquid.ts [hunk 1]
```diff
@@ -161,6 +188,107 @@ export default class hyperliquid extends hyperliquidRest {
...
+            parsedTickers.push (ticker);
+        }
+        // perpetuals
+        const meta = this.safeDict (data, 'meta', {});
+        const universe = this.safeList (meta, 'universe', []);
+        const assets = this.safeList (data, 'assetCtxs', []);
+        for (let i = 0; i < assets.length; i++) {
+            let assetObject = assets[i];
+            const coinObject = universe[i];
+            const id = coinObject['name'] + ':USDC/USDC';
```
Comment (carlosmiei) [inline ts/src/pro/hyperliquid.ts:254]: @ttodua this is not correct, we're injecting as an exchange-specific field `coin` then `info` contains this field

```
'coin': 'kSHIB:USDC/USDC',
                              'dayNtlVlm': '2299289.598427',
                              'funding': '-0.00000479',
                              'impactPxs': ['0.019278', '0.019279']
```

File: ts/src/pro/hyperliquid.ts [hunk 2]
```diff
@@ -161,6 +188,107 @@ export default class hyperliquid extends hyperliquidRest {
...
+            assetObject = this.extend (assetObject, { 'coin': id });
+            const ticker = this.parseWsTicker (assetObject, 'swap');
+            parsedTickers.push (ticker);
+        }
+        const tickers = this.indexBy (parsedTickers, 'symbol');
+        client.resolve (tickers, 'tickers');
+    }
+
+    parseWsTicker (message, marketType) {
```
Comment (carlosmiei) [inline ts/src/pro/hyperliquid.ts:263]: also we shouldn't change this signature

Comment (ttodua) [reply ts/src/pro/hyperliquid.ts:263]: >also we shouldn't change this signature 

 but binance has that signature. right, it's not correct signature, i had blindly copied from binance and this also means binance also needs fix

File: ts/src/pro/hyperliquid.ts (hunk 1 above)
Comment (ttodua) [reply ts/src/pro/hyperliquid.ts:254]: good catch, i missed that

**Comment by ttodua** [comment]
at first, https://github.com/ccxt/ccxt/pull/23117 should be merged, then review this current PR

**Comment by carlosmiei** [comment]
@ttodua we're missing some skips to make the build pass, can you check it?

**Comment by ttodua** [comment]
sorry, i was on the rush in the morning and couldnt finish. i'll turn into ready and tag you soon.

File: ts/src/pro/hyperliquid.ts [hunk 3]
```diff
@@ -125,6 +125,33 @@ export default class hyperliquid extends hyperliquidRest {
...
+        const url = this.urls['api']['ws']['public'];
+        const request: Dict = {
+            'method': 'subscribe',
+            'subscription': {
+                'type': 'webData2', // allMids
+                'user': '0x0000000000000000000000000000000000000000',
+            },
+        };
+        const tickers = await this.watch (url, messageHash, this.extend (request, params), messageHash);
+        if (this.newUpdates) {
```
[the remark this answers is not in the document]
Comment (carlosmiei) [reply ts/src/pro/hyperliquid.ts:149]: @ttodua the symbol here is wrong hence the result I was getting 

image

**Comment by carlosmiei** [comment]
```
p hyperliquid watchTickers '["BTC/USDC:USDC", "PURR/USDC"]'
Python v3.12.3
CCXT v4.3.65
hyperliquid.watchTickers(['BTC/USDC:USDC', 'PURR/USDC'])
{'BTC/USDC:USDC': {'ask': 67449.0,
                   'askVolume': None,
                   'average': None,
                   'baseVolume': None,
                   'bid': 67448.0,
                   'bidVolume': None,
                   'change': None,
                   'close': 67448.5,
                   'datetime': None,
                   'high': None,
                   'info': {'dayNtlVlm': '337583734.06514001',
                            'funding': '0.0000125',
                            'impactPxs': ['67448.0', '67449.0'],
                            'markPx': '67443.0',
                            'maxLeverage': 50,
                            'midPx': '67448.5',
                            'name': 'BTC',
                            'onlyIsolated': False,
                            'openInterest': '2262.38718',
                            'oraclePx': '67410.0',
                            'premium': '0.00056371',
                            'prevDayPx': '67043.0',
                            'szDecimals': 5},
                   'last': 67448.5,
                   'low': None,
                   'open': None,
                   'percentage': None,
                   'previousClose': 67043.0,
                   'quoteVolume': 337583734.06514,
                   'symbol': 'BTC/USDC:USDC',
                   'timestamp': None,
                   'vwap': None},
 'PURR/USDC': {'ask': None,
               'askVolume': None,
               'average': None,
               'baseVolume': None,
               'bid': None,
               'bidVolume': None,
               'change': None,
               'close': 0.21264,
               'datetime': None,
               'high': None,
               'info': {'circulatingSupply': '598724619.56353998',
                        'coin': 'PURR/USDC',
                        'dayNtlVlm': '7331928.07087',
                        'markPx': '0.21265',
                        'midPx': '0.21264',
                        'prevDayPx': '0.19778'},
               'last': 0.21264,
               'low': None,
               'open': None,
               'percentage': None,
               'previousClose': 0.19778,
               'quoteVolume': 7331928.07087,
               'symbol': 'PURR/USDC',
               'timestamp': None,
               'vwap': None}}
```
```
p hyperliquid fetchTickers '["BTC/USDC:USDC", "PURR/USDC"]'
Python v3.12.3
CCXT v4.3.65
hyperliquid.fetchTickers(['BTC/USDC:USDC', 'PURR/USDC'])
{'BTC/USDC:USDC': {'ask': None,
                   'askVolume': None,
                   'average': None,
                   'baseVolume': None,
                   'bid': None,
                   'bidVolume': None,
                   'change': None,
                   'close': None,
                   'datetime': None,
                   'high': None,
                   'info': {'baseId': 0,
                            'maxLeverage': '50',
                            'name': 'BTC',
                            'onlyIsolated': False,
                            'szDecimals': '5'},
                   'last': None,
                   'low': None,
                   'open': None,
                   'percentage': None,
                   'previousClose': None,
                   'quoteVolume': None,
                   'symbol': 'BTC/USDC:USDC',
                   'timestamp': None,
                   'vwap': None},
 'PURR/USDC': {'ask': None,
               'askVolume': None,
               'average': None,
               'baseVolume': None,
               'bid': None,
               'bidVolume': None,
               'change': None,
               'close': 0.212625,
               'datetime': None,
               'high': None,
               'info': {'circulatingSupply': '598724619.38378',
                        'coin': 'PURR/USDC',
                        'dayNtlVlm': '7332154.11692',
                        'index': '0',
                        'isCanonical': True,
                        'markPx': '0.21262',
                        'midPx': '0.212625',
                        'name': 'PURR/USDC',
                        'prevDayPx': '0.19778',
                        'tokens': ['1', '0']},
               'last': 0.212625,
               'low': None,
               'open': None,
               'percentage': None,
               'previousClose': 0.19778,
               'quoteVolume': 7332154.11692,
               'symbol': 'PURR/USDC',
               'timestamp': None,
               'vwap': None}}
```

File: ts/src/pro/hyperliquid.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ttodua) [reply ts/src/pro/hyperliquid.ts:149]: > @ttodua the symbol here is wrong hence the result I was getting
> 
> image

🤦

**Pull request merged**